### PR TITLE
[JIT] Add support for named_modules()

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29495 [JIT] Add support for named_modules()**

Fix for https://github.com/pytorch/pytorch/issues/28998

Differential Revision: [D18412561](https://our.internmc.facebook.com/intern/diff/D18412561)

### CODE REVIEW

File: torch/csrc/jit/script/python_sugared_value.h [hunk 1]
```diff
@@ -102,6 +102,70 @@ struct VISIBILITY_HIDDEN ConstantParameterList : public SugaredValue {
...
+      return std::make_shared<ModuleDictMethod>(keys_, "keys");
+    } else if (field == "values") {
+      return std::make_shared<ModuleDictMethod>(modules_, "values");
+    } else if (field == "items") {
+      auto iterator = std::make_shared<IterableTree>();
+      iterator->addChild(loc, m, keys_);
+      iterator->addChild(loc, m, modules_);
+      return std::make_shared<ModuleDictMethod>(iterator, "items");
+    } else {
+      throw ErrorReport(loc) << "_Modules only supports keys, values, or items";
```
Comment (driazati) [inline torch/csrc/jit/script/python_sugared_value.h]: `_Modules` -> `_modules`

**Review by driazati — APPROVED** [review]

**Review by zdevito — DISMISSED** [review]
Binding a private python member into TorchScript seems like the wrong approach. Shouldn't we implement the standard nn.Module API for this stuff? That would be the `children`, `named_children`, `modules()` and `named_modules` methods. Having `_modules` without `named_children` is going to promote the use of a private API.

**Comment by eellison** [comment]
This doesn't do the checking for uniqueness, but when we recursively script we create a new module on each compilation (don't preserve aliasing), so i don't think that's actually a problem.

File: torch/csrc/jit/script/python_sugared_value.h [hunk 2]
```diff
@@ -144,40 +175,71 @@ struct VISIBILITY_HIDDEN ModuleValue : public SugaredValue {
...
-      at::ArrayRef<NamedValue> inputs,
-      at::ArrayRef<NamedValue> attributes,
-      size_t n_binders) override {
-    if (inputs.size() || attributes.size()) {
-      throw ErrorReport(loc)
-          << name_ << " method does not accept any arguments";
-    }
-    return iterable_;
+      Function& m,
+      const std::string& field) override {
```
Comment (driazati) [inline torch/csrc/jit/script/python_sugared_value.h]: Can you move the implementation to the `.cpp` file

File: torch/csrc/jit/script/python_sugared_value.cpp [hunk 3]
```diff
@@ -316,24 +334,16 @@ std::shared_ptr<SugaredValue> ModuleValue::attr(
...
-        get_keys = true;
-        get_values = true;
-      } else if (field == "values") {
-        get_values = true;
-      } else {
-        get_keys = true;
-      }
-      return std::make_shared<ModuleDictMethod>(
-          desugarModuleContainer(get_keys, get_values, loc, m), field);
+    if (field == "items" || field == "keys" || field != "values") {
```
Comment (driazati) [inline torch/csrc/jit/script/python_sugared_value.cpp]: This should be `field == "values"` right

**Review by driazati — APPROVED** [review]
Can you make a follow up issue or add to this PR stuff like `named_children`?

**Comment by eellison** [comment]
> Can you make a follow up issue or add to this PR stuff like `named_children`?

i'm going to do the other apis as a follow up sometime this week.

**Comment by eellison** [comment]
@chenbohua3  thx !

**Pull request closed without merging**